course/latex/latex.py

```diff
@@ -0,0 +1,176 @@
+# -*- coding: utf-8 -*-
+
+from __future__ import division
+
+__copyright__ = "Copyright (C) 2016 Dong Zhuang, Andreas Kloeckner"
+
+__license__ = """
+Permission is hereby granted, free of charge, to any person obtaining a copy
+of this software and associated documentation files (the "Software"), to deal
+in the Software without restriction, including without limitation the rights
+to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+copies of the Software, and to permit persons to whom the Software is
+furnished to do so, subject to the following conditions:
+
+The above copyright notice and this permission notice shall be included in
+all copies or substantial portions of the Software.
+
+THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+THE SOFTWARE.
+"""
+
+import re
+
+from django.utils.translation import ugettext as _
+
+from .utils import strip_comments, strip_spaces
+
+class TexDocParseError(Exception):
+    pass
+
+class TexDocMissingElementError(TexDocParseError):
+    pass
+
+class TexDocWrongElementOrderError(TexDocParseError):
+    pass
+
+class TexDoc():
+    """
+    Defines a LaTeX document
+    """
+    preamble = ""
+    document = ""
+    has_preamble = False
+    has_begindoc = False
+    has_enddoc = False
+
+    def is_empty_pagestyle_already(self):
+        match = re.search(r"\\pagestyle{\s?empty\s?}", self.preamble)
+        if match:
+            return True
+        return False
+
+    def parse(self, latex, test=False):
```

Comment:
With this, we can call the macro like [this](https://github.com/inducer/relate-sample/pull/4/files#diff-0bdbb3e999a8bd689c5cff408129ecb8R53). If not, ``parse`` will fail because there are no preamble in the call block. Because a full latex doc should contain ``\documentclass``, ``\begin{document}``, ``\end{document}``.

``parse`` using ``test=True`` (which might need to be renamed so as not to cause misunderstanding) is to check whether preamble are defined out side the block, and assemble them into full latex code, and try parse again.